src/execution/futex-emulation.cc

```diff
@@ -330,6 +348,13 @@ Object FutexEmulation::Wait(Isolate* isolate, WaitMode mode,
       rel_timeout_ns = static_cast<int64_t>(timeout_ns);
     }
   }
+
+  // Fence after `Wait` on ordered threads.
+  int id = GetRecordReplayWaitOrderedLock();
+  base::Optional<recordreplay::AutoOrderedLock> ordered;
+  if (!recordreplay::AreEventsDisallowed()) {
+    ordered.emplace(id);
+  }
```

Comment:
I think this change and the associated ordered lock should be removed.  AFAICT the issue doesn't have an ordering problem around the futex accesses, so this is fixing a problem we don't have, and this fix has some problems:

* There's no usage of this ordered lock during notification, which is needed to ensure threads using these futexes to communicate will actually run in a consistent order.
* Wrapping a wait for an indefinite amount of time with an ordered lock means that the ordered lock will be held throughout the wait and can lead to deadlocks.  Ordering notify/wakeup calls is tricky and the best thing to do is use an ordered e.g. pthread_mutex_t and the system library methods for notifying / waking things up.

Ordering uses of JS Atomics isn't a priority.  My understanding is that these are primarily for use with SharedArrayBuffer, which I think we allow to be used but which we don't want to replay with exact fidelity -- we don't record/replay accesses from different threads in the same order and this is a case where if we can replay it, great, but if we can't there are other things we should work on fixing instead.

Reply:
Thanks for the feedback!

1. **Non-deterministic JS:** I guess my concern was that checkpoints could diverge if the timing of certain JS instructions diverged. `Atomic.wait` and `Atomic.asyncWait` do put the thread to sleep, so if they come back at different times, this might be an issue. We can just roll it out and see if it causes problems, but I still feel like we need to add ordering here.
2. **Ordered lock**: Whoops! The ordered lock was used incorrectly. We should only sync *after* wait (and maybe before, but that's not too important), not during.
3. **Unordered `notify`**: We can also add ordering on `notify`, I agree, but I feel that the main source of divergence can come from the fact that an ordered thread depends on an unordered thread to notify it. In that case, syncing after `Wait` would fix that problem.

Sure, we can ship without the lock. I'll put an `Assert` in after `Wait`, so that we know if unordered threads ever notify ordered threads.